### DISCUSSION TITLE
Automatically patch `js.document` and `js.window` in Workers

### DISCUSSION THREAD

**Discussion by WebReflection** [pyscript/pyscript#1878] [Proposals]
Not only some Pyodide module assumes by default that code runs on the main thread, hence it reaches `js.document` without thinking that could be absent, we once again had some user finding the lack of `document` problematic.

I remember me and @fpliger also needed to manually patch that and point at the proxied version and I start wondering, as we already offer a `RUNNING_IN_WORKER` constant via `pyscript` module, if we shouldn't just stretch further and enable more modules and reduce surprises when a script `worker` attribute is used.

That means that beside reaching out for `pyscript.window` and `pyscript.document`, the global Worker context gets polluted with both references pointing at our module.

This makes the code easier to migrate but I am not sure this might create confusion, as w estill want people to understand that `pyscript` is the module they should use.

However, we don't get to control 3rd party packages and if these assume, behind their scene, that `js.document` exists, we might avoid breaking out of the blue by monkey-patching that on our side *after* we are sure we're running in a worker.

@ntoll @JeffersGlass @antocuni @madhur-tandon @leriomaggio ... thoughts?

**Comment by fpliger** [comment]
This is a great topic. Thanks @WebReflection for kicking it off.

Agree with you that we have a little bit of a dilemma having to choose transparency versus compatibility/usability. All In all, I think we should lean towards making the user experience as pleasant as possible while at the same time leaving an escape hatch.

How about we keep the patches as a default behavior and offer a config option to turn it off? This way the default will probably match what most users would need but we still have an option for people to explicitly turn it off and use Pyodide's own interface.

**Reply by WebReflection** [reply]
thing is ... there's no Pyodide own interface for `document` when running in a Worker and I think `del js.document` is all they need instead of another config thing to learn and parse ... wouldn't that be equivalent if we want to provide an escape hatch for whatever reason? My main concern is that *some* package *might* actually do the same check we do to understand if `document` exist ... but I am not sure how to validate this theory

**Reply by ntoll** [reply]
My intuition is the same as Fabio's.

I'm also concerned that we could over-think this, when "that's just how the universe works" is perhaps the best answer we can give. We can't possibly deal with every possible eventuality (e.g. a package already checking for a `document` already). That's not our responsibility I guess. Hence the need for escape hatches.

**Reply by WebReflection** [reply]
how would we call that escape hatch? `patch_worker = false` and we assume `true` by default?

**edit** actually `patch_worker_dom = false` would be even more explicit but I am open to suggestions

**Reply by ntoll** [reply]
Yeah, by escape hatch I mean something like the flag you suggest.

**Reply by WebReflection** [reply]
something to overthink about too ... as we can't predict the whole ecosystem, there's also the case one package expects `js.document` to be reachable and another one instead checking for `document` attribute in `js` as we do already to know if it's a Worker or not ... this disambiguation in the config, as discussed as escape hatch, won't be able to solve that case ... this is why I'd love to have data instead of packages assuming `js.document` exists VS packages branching logic if that's not the case (although if a package assumes `js.document` what is it going to do when such reference doesn't exist?).

In short, for users sake I'd say *patch all the things* and make code portable across main/worker threads ... but if it's escaping that we're after then we need to think more about how to hint the config that package in particular is the only one that doesn't want a `document`, and stuff becomes way more complicated rather than being easier on users, but I also see this use case very rare and, whenever is needed, good enough to explain as long as we can target explicit packages and not the patch globally. I've also no idea how to do that on package basis so if we provide complications that don't solve really complicated scenarios, I'd be rather back to make it a no-brainer default and let anyone `del js.document` when that's really not desired within a worker.

Let's also face reality that neither pyodide nor MicroPython can deal with a document if running in a worker so that I think we should be good just patching it and call it a day.

**Reply by JeffersGlass** [reply]
So by default we would, only in the worker context, do `js.document = pyscript.document`, and if that flag is set to false, we just don’t do this?

I like this idea, as I think it will help more modules work out of the box (Matplotlib being the one that leaps to mind).

**Reply by WebReflection** [reply]
@JeffersGlass yes, the idea is that one exactly but the more I think about it the more the escape in config doesn't work ... because it disables matplotlib out of the box because just one package did something weird and unexpected, with the current state of both Pyodide or MicroPython, to date.

To me this should always work:

```python
import js
js.document.body.textContent = 'Hello';
```

as that's kinda the de-facto standard expectation for all pyodide modules to date ... however, it's still possible to *escape* in code:

```python
import js
document = js.document

# let the exception do its thing
del js.document
import weird_package_that_branches_logic

# re-enable all others
js.document = document
```

This is declarative and explicit and it's easier to reason about, if ever needed, rather than complicating the config with logic that deals with only that weird package and that, at the end of the day, would do exactly the same ... hoping also that weird package won't try to disambiguate worker VS non worker at runtime or there's literally no solution to satisfy all cases.

**Reply by JeffersGlass** [reply]
This makes sense to me! I vote we patch and be done, no flag necessary - and if someone comes along with a reeeeally convincing case for why a global patch flag would be useful, we can always add one later. 👍

**Reply by ntoll** [reply]
OK... just catching up. LGTM :shipit:

**Reply by WebReflection** [reply]
truth to be told, and for argument sake, we could also really create this *Troubleshooting* entry in our docs and suggest:

```python
from pyscript import document
import js
js.document = document
```

and call it a day ... that would do nothing in the main thread, it will solve various packages in the worker case ... so maybe all we lack is not more magic, rather a section to solve common issues in our docs?

**Reply by JeffersGlass** [reply]
Hmmmm…. More and more I hear that people don’t read the documentation 😇.

I’d be game for taking the slightly-more-aggressive approach and doing the patch.

The middle ground could be to throw an error when the user tries to access `js.document` in a worker and direct them to our docs… but at that point, I think just doing the patch for them is still better.

**Reply by WebReflection** [reply]
last thoughts from me:

> I vote we patch and be done, no flag necessary

if the config flag is to opt-out, we could indeed just make the patch the default and re-discuss, and re-consider, the config flag in the future, whenever anyone will have issues with us patching globally the `document` ... this is, imho, a great way forward without making us stuck in complications that might never be needed ... to date, all the info we have is that people don't expect packages to break, matplotlib being probably the most prominent example.

**Reply by WebReflection** [reply]
we're here :partying_face: https://github.com/pyscript/pyscript/pull/1880

**Reply by WebReflection** [reply]
this is for @fpliger ... we all seem to agree with the current MR and willing to re-iterate later on if the use case for disabling it *per package* makes sense (as opposite of globally, as that breaks other modules) ... is that OK to you?

The MR is already approved: https://github.com/pyscript/pyscript/pull/1880 I just want to be sure we're all aligned in here, thanks!

**Reply by fpliger** [reply]
That works for me too! Looking at the PR, it's only adding the patch and no disabling at all, right? 

Asking to explicitly confirm the intentions not for an issue per se. I think it's fine as is since js would mostly not work on a worker anyway (I think... It's early morning 😅)

**Reply by WebReflection** [reply]
correct, the PR just brings the document in so that matplotlib among other packages would just work from a worker, but of course if extra weird stuff is needed we can refine or rethink the patch
